### DISCUSSION TITLE
Package possibly not reading environment variables correctly

### DISCUSSION THREAD

**Discussion by william-chu-github** [oracle/python-oracledb#436] [Q&A] [answered]
I honestly don't know if the problem is with `python-oracledb` or something else, as I will outline below. The gist of it is that I have some code that needs to connect to an Oracle database, and it is not picking up the path environment variable. This means I have to initialise with `oracledb.init_oracle_client(lib_dir = ...)`, which causes maintenance issues/problems with less technical users. Here's an annoted illustration of my experience.

1. Start a new venv with `python -m venv test`
2. `test\scripts\activate` to activate the venv
3. `pip install oracledb`
4. Start a Python session in a command window
```
import oracledb
oracledb.init_oracle_client()
```
5. Everything works fine

Windows 11 Enterprise version 23H2, Python 3.10.11, oracle-db 2.5.1.

Now, I move over to my IDE, Spyder.

1. In a command prompt, `set DPI_DEBUG_LEVEL=64` to set up tracing
2. Start and set Spyder to use the new venv
3. Try to initialise the client as before, which errors with this tracing information (notice it doesn't show the full path from the environment variable; just a subdirectory in a user folder):
```
... Environment Variables:
...     PATH => %USERPROFILE%\AppData\Local\Microsoft\WindowsApps;"
```
4. But `os.environ["PATH"]` explicitly includes the directory to the necessary configuration files!
5. Mind-blowingly, if I do `os.environ["PATH"] += ";<path_to_oci.dll>"`, then try `oracledb.init_oracle_client()`--it now works. The trace will show the full environment path (even down to the duplicated Oracle directory because I did a `+=` to append it when it was already in the path).

Does anyone have any idea what's going on here? Is it a problem with `oracledb` or perhaps Spyder? The fact that it works in a plain command prompt would indicate that `oracledb` is working correctly, but then it's not picking up the environment path correctly, when Spyder is.

Thanks for reading.

**Comment by anthony-tuininga** [comment]
Since oracledb is calling `getenv()` and not the Windows specific `GetEnvironmentVariable()` method, that may explain the difference? I don't use Windows generally and have never used Spyder so I don't know what it is doing internally -- and whether there is a discrepancy between `getenv()` and `GetEnvironmentVariable()` on Windows in this particular case. If you find out for sure, please advise!

**Reply by anthony-tuininga** [reply, marked as answer]
It seems that is indeed the case based on some Google searches: https://github.com/curl/curl/issues/4774. Apparently Windows' implementation of `getenv()` is .... lacking... :-(

**Reply by william-chu-github** [reply]
Thanks, that might be it. I wasn't able to verify it because I couldn't find a likely spot where `getenv()` was called (a search of the repo only found it in [connect_params.pyx](https://github.com/oracle/python-oracledb/blob/eb1daa48a8f26e927e51d36b4a733e55ffbe25ef/src/oracledb/impl/base/connect_params.pyx#L58)), and at any rate, I think IT blocks us from pulling from GitHub.

I can work around it just be trivially setting `os.environ["PATH"] = os.environ["PATH"]`. Perhaps it might be helpful for someone to test what OS is running and conditionally use `GetEnvironmentVariable()` if it's Windows?

**Reply by anthony-tuininga** [reply]
FYI, changes to use `SetEnvironment()` on Windows were made in the latest commits to python-oracledb. You can grab the latest build and test if you have a mind to do so! Or wait until 3.0 is released.

**Reply by anthony-tuininga** [reply]
And 3.0.0 has been released with this fix included. If you can confirm that the issue is resolved for you, that would be appreciated!

**Reply by william-chu-github** [reply]
Yep, 3.0.0 seems to have resolved this. Thanks!
